### PR TITLE
[Android] Brave news - ui fixes

### PR DESCRIPTION
**Pull request by alexsafe**
Resolves https://github.com/brave/brave-browser/issues/21868
Resolves https://github.com/brave/brave-browser/issues/20580
Resolves https://github.com/brave/brave-browser/issues/21873

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by tapanmodh — APPROVED** [review]
LGTM

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java [hunk 1]
```diff
@@ -264,7 +264,7 @@ protected void onFinishInflate() {
             mComesFromNewTab = BraveActivity.getBraveActivity().isComesFromNewTab();
             mIsFeedLoaded = BraveActivity.getBraveActivity().isLoadedFeed();
             mExistingNewsFeedObject = BraveActivity.getBraveActivity().getNewsItemsFeedCards();
-            mPrevScrollPosition = 1;
+            mPrevScrollPosition = -1;
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java:267]: maybe it worth to replace `-1`, `-2`  with self-explaining named constants?

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java [hunk 2]
```diff
@@ -705,6 +727,57 @@ private void refreshFeed() {
         }
     }
 
+    private void keepPosition(int prevScrollPosition) {
+        processFeed();
+        int scrollPosition = prevScrollPosition;
+
+        if (scrollPosition == -2) {
+            if (ConfigurationUtils.isLandscape(mActivity)) {
+                new Handler().postDelayed(new Runnable() {
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java:736]: nit: you may consider to use more compact syntax
```
                        new Handler().postDelayed(() -> {
                            if (mRecyclerView != null) {
                                mRecyclerView.scrollToPosition(0);
                            }
                         }, 100);

```
here and in other places

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java [hunk 3]
```diff
@@ -1096,7 +1154,30 @@ public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newStat
...
+                                                    && scrollY > 600)
+                                                || (!ConfigurationUtils.isLandscape(mActivity)
+                                                        && scrollY > 900))) {
+                                    position = -2;
+                                } else {
+                                    position = firstCompletelyVisibleItemPosition == 0
+                                            ? -1
+                                            : firstCompletelyVisibleItemPosition;
+                                }
+                                if (tab != null) {
```
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java:1175]: this is always true, as you have `&& BraveActivity.getBraveActivity().getActivityTab() != null)` above

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java (hunk 1 above)
Comment (alexsafe) [reply android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java]: @AlexeyBarabash good point. replaced

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java [hunk 4]
```diff
@@ -1096,7 +1154,30 @@ public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newStat
                                 linearLayoutManager.findFirstVisibleItemPosition();
                         int lastVisibleItemPosition =
                                 linearLayoutManager.findLastVisibleItemPosition();
-
+                        int scrollY = mParentScrollView.getScrollY();
+
+                        if (BraveActivity.getBraveActivity() != null
+                                && BraveActivity.getBraveActivity().getActivityTab() != null) {
+                            Tab tab = BraveActivity.getBraveActivity().getActivityTab();
```
[the remark this answers is not in the document]
Comment (alexsafe) [reply android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java]: refactored

File: android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java (hunk 3 above)
Comment (alexsafe) [reply android/java/org/chromium/chrome/browser/ntp/BraveNewTabPageLayout.java]: refactored

**Review by SergeyZhukovsky — APPROVED** [review]
++

**Review by AlexeyBarabash — APPROVED** [review]
thanks :+1:

**Comment by kjozwiak** [comment]
#### `Test Case #1` - https://github.com/brave/brave-browser/issues/21868

Verification **PASSED** on `Samsung Galaxy Tablet A` running `Android 12` using the following build(s):

```
1.39.63 Chromium: 100.0.4896.127
```

Went through the STR/Cases outlined via https://github.com/brave/brave-browser/issues/21868#issue-1180108509 and ensured that switching the orientation didn't cause the original issues that @Uni-verse originally reported.

Example of the fix/issue not occurring with the latest Nightly.

https://user-images.githubusercontent.com/2602313/163657865-2ab1b6db-be34-412d-8c5a-18482a5a8af7.mp4

#### `Test Case #2` - https://github.com/brave/brave-browser/issues/20580

Regarding the STR/Cases mentioned https://github.com/brave/brave-browser/issues/20580#issue-1108745502, there still seems to be some issues with the Brave News jumping around but it has definitely improved compared to before. Quick example:

https://user-images.githubusercontent.com/2602313/163659019-500dde25-14eb-4881-ac4a-1ea844859cd4.mp4

Went through the STR/Cases outlined via https://github.com/brave/brave-browser/issues/20580#issuecomment-1024125578 and ensured that the Brave News wasn't jumping around when refreshing a NTP. Also ensured that when refreshing a NTP, the feed restarts and starts at the top again.

https://user-images.githubusercontent.com/2602313/163658736-a17f20fc-8ab6-4431-b271-786efe43640b.mp4

#### `Test Case #3` - https://github.com/brave/brave-browser/issues/21873

Went through the STR/Cases oulined via https://github.com/brave/brave-browser/issues/21873#issue-1180343446 and ensured that Brave News wasn't flickering by appearing like it was loading and then moved to the bottom. There's still some flickering which I think is due to  #22187.

Example of the original issue not happening anymore:

https://user-images.githubusercontent.com/2602313/163658479-ad044935-3b20-445a-b229-968dee8bfbf0.mp4

**Comment by kjozwiak** [comment]
Created a follow up via https://github.com/brave/brave-browser/issues/22356 to improve Android keeping the correct position with the Brave News feed when switching tabs.

**Pull request merged**